CASMINST-5609: Refine and clarify NCN personalization procedures (#2853)
Cray-HPE/docs-csm: operations/CSM_product_management/Configure_Non-Compute_Nodes_with_CFS.md

the [Configuration Framework Service (CFS)](../configuration_management/Configuration_Management.md).

During a fresh install, carry out these procedures in order. Later, individual
procedures may be re-run as needed. These procedures are not done as part of CSM upgrades.

> In particular, during upgrades, different procedures are provided that handle NCN

Author: ```suggestion
In particular, during upgrades, different procedures are provided that handle management NCN
```

   - [Option 1: Automated default](#option-1-automated-default)
   - [Option 2: Manual](#option-2-manual)

> 3. [Perform NCN personalization](#3-perform-ncn-personalization)

Author: ```suggestion
3. [Perform Management NCN personalization](#3-perform-management-ncn-personalization)
```

   - [Option 1: Automatic configuration](#option-1-automatic-configuration)

>      - [Automatic configuration: CSM install](#automatic-configuration-csm-install)

Author: ```suggestion
     - [Automatic configuration: CSM fresh install](#automatic-configuration-csm-fresh-install)
```

## 3. Perform NCN personalization

> > If a CSM upgrade is being performed, then this procedure is only applicable if it is a CSM-only upgrade.

Author: That's right. In the past we did sometimes direct users here during upgrades, but I think you are correct that as of csm-1.3 that is no longer the case. So probably I should edit this to reflect that.

> That is, an upgrade on a system with no other products installed. **This is extremely uncommon in production
> environments**. For upgrades that are not CSM-only, a different procedure is provided for handling NCN
> personalization. Review the upgrade documentation for more information.

The previous procedures on this page did not make any changes to the NCNs on the system. They merely
specified how the NCNs should be configured. In order for these configurations to be applied to the NCNs,
a process called NCN personalization takes place using [CFS](../configuration_management/Configuration_Management.md).

There are two necessary steps:

1. Create or update the NCN personalization configuration in CFS.
1. Update CFS to set the NCN personalization configuration as the desired configuration for the NCNs.

> These can be accomplished by [running the `apply_csm_configuration.sh` script](#option-1-automatically-apply-csm-configuration),

Author: I agree with you. I've made the change.

  1. Finds all nodes in HSM with the `Management` role.
  1. Disables all management NCNs in CFS.
  1. Updates the `ncn-personalization` configuration in CFS to contain only the CSM layer for the latest installed CSM release.
  1. In CFS, enables all management NCNs, clears their error count, and sets their desired configuration to `ncn-personalization`.
  1. Monitors CFS until all management NCNs have successfully completed or failed configuration.

- In its other mode (specified by using the `--no-config-change`), the script applies an existing CFS configuration to
  the management NCNs. In this mode, the script performs the following steps:

  1. Finds all nodes in HSM with the `Management` role.
  1. In CFS, enables all management NCNs, clears their error count, and sets their desired configuration to `ncn-personalization`.
  1. Monitors CFS until all management NCNs have successfully completed or failed configuration.

#### Automatic configuration: Usage

(`ncn-mw#`) View a usage message for the script, documenting all of its parameters, by running the following command:

```bash

> /usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh

Author: ```suggestion
/usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh -h
```

> ## 3. Perform NCN personalization

Author: ```suggestion
## 3. Perform Management NCN personalization
```

The previous procedures on this page did not make any changes to the NCNs on the system. They merely
specified how the NCNs should be configured. In order for these configurations to be applied to the NCNs,
a process called NCN personalization takes place using [CFS](../configuration_management/Configuration_Management.md).

There are two steps:

1. Create or update the NCN personalization configuration in CFS, if needed.
1. Update CFS to set the NCN personalization configuration as the desired configuration for the NCNs.

These can be accomplished by following [Option 1: Automatic configuration](#option-1-automatic-configuration) or [Option 2: Manual configuration](#option-2-manual-configuration).

### Option 1: Automatic configuration

This option uses a script to create or update the NCN personalization configuration in CFS, and then to
apply the configuration to the NCNs.

> The `docs-csm` RPM must be installed in order to use this script. See
> [Check for Latest Documentation](../../update_product_stream/README.md#check-for-latest-documentation).

The options used with the script depend on the context in which this procedure is being followed.

> - If this procedure is being run for the first time, during a CSM install, then follow the

Author: ```suggestion
- If this procedure is being run for the first time, during a CSM fresh install, then follow the
```

>   [Automatic configuration: CSM install](#automatic-configuration-csm-install) procedure.

Author: ```suggestion
  [Automatic configuration: CSM fresh install](#automatic-configuration-csm-fresh-install) procedure.
```

- If this procedure is being run later in order to make changes to passwordless SSH, root user password, or root user SSH keys,
  then follow the [Automatic configuration: Apply changes](#automatic-configuration-apply-changes) procedure. That is also the
  procedure to follow in order to force NCN personalization to be re-run on all of the NCNs.

In either case, although it is not usually required, it is possible to further customize the behavior of the script.
For more information, see:

- [Automatic configuration: Modes of operation](#automatic-configuration-modes-of-operation).
- [Automatic configuration: Usage](#automatic-configuration-usage).
- [Automatic configuration: Parameters](#automatic-configuration-parameters).

> #### Automatic configuration: CSM install

Author: ```suggestion
#### Automatic configuration: CSM fresh install
```

> > At this point in the installation, the only product installed on the system is CSM. Therefore the CFS configuration

Author: ```suggestion
> At this point in the fresh install process, the only product installed on the system is CSM. Therefore the CFS configuration
```

> > that the script creates for NCN personalization will only contain the CSM layer.

Author: ```suggestion
> that the script creates for management NCN personalization will only contain the CSM layer.
```

(`ncn-mw#`) When performing this procedure during a CSM fresh install, the recommended command to run is the following:

```bash
/usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh
```

#### Automatic configuration: Apply changes

> (`ncn-mw#`) In order to force NCN personalization to run on all of the NCNs, run the following command:

Author: ```suggestion
(`ncn-mw#`) In order to force NCN personalization to run on all of the management NCNs, run the following command:
```

```bash
/usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh --no-config-change --clear-state
```

#### Automatic configuration: Modes of operation

The script has two basic modes of operation:

- In the default mode (which can be explicitly specified by using the

>   `--config-change` parameter), the script will create or update a CFS configuration, and then apply it to the NCNs.

Author: ```suggestion
  `--config-change` parameter), the script will create or update a CFS configuration, and then apply it to the management NCNs.
```

  In this mode, the script performs the following steps:

  1. Finds the latest installed release version of the CSM product stream.
  1. Finds the CSM configuration version associated with the given release.
  1. Finds the latest commit on the corresponding release branch of the `csm-config-management` repository.

>   1. Creates or updates the `ncn-personalization.json` configuration file with the new CSM commit.

Author: ```suggestion
```

  1. Finds all nodes in HSM with the `Management` role.

>   1. Disables configuration for all NCNs.

Author: ```suggestion
  1. Disables configuration for all management NCNs.
```

>   1. Updates the `ncn-personalization` configuration in CFS from the `ncn-personalization.json` file.

Author: ```suggestion
  1. Updates the `ncn-personalization` configuration in CFS.
```

>   1. Enables configuration for all NCN nodes, and sets their desired configuration to `ncn-personalization`.

Author: ```suggestion
  1. Enables configuration for all management NCNs, and sets their desired configuration to `ncn-personalization`.
```

>   1. Monitors CFS until all NCN nodes have successfully completed or failed configuration.

Author: ```suggestion
  1. Monitors CFS until all management NCNs have successfully completed or failed configuration.
```

- In its other mode (specified by using the `--no-config-change`), the script applies an existing CFS configuration to

>   the NCNs. In this mode, the script performs the following steps:

Author: ```suggestion
  the management NCNs. In this mode, the script performs the following steps:
```

  1. Finds all nodes in HSM with the `Management` role.

>   1. Enables configuration for all NCN nodes, and sets their desired configuration to `ncn-personalization`.

Author: ```suggestion
  1. Enables configuration for all management NCNs, and sets their desired configuration to `ncn-personalization`.
```

>   1. Monitors CFS until all NCN nodes have successfully completed or failed configuration.

Author: ```suggestion
  1. Monitors CFS until all management NCNs have successfully completed or failed configuration.
```

#### Automatic configuration: Usage

(`ncn-mw#`) View a usage message for the script, documenting all of its parameters, by running the following command:

```bash
/usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh
```

#### Automatic configuration: Parameters

The script also supports several flags to override the default behaviors described previously.
Some of these flags apply to both modes of operation, whereas others apply only to the
`--config-change` mode.

- Common parameters
  - `--clear-state`: Clears existing state from components to ensure that CFS runs, even if no
    changes have been made to the content of the CFS configuration.
  - `--config-name`: For `--config-change` mode, this is the name of the CFS configuration that
    is created or updated. For `--no-config-change` mode, this is the name of the existing CFS configuration

>     to use on the NCNs. In either case, it defaults to `ncn-personalization`.

Author: ```suggestion
    to use on the management NCNs. In either case, it defaults to `ncn-personalization`.
```

>   - `--no-clear-err`: By default, the error count is cleared on all of the NCN CFS components. If this

Author: ```suggestion
  - `--no-clear-err`: By default, the error count is cleared on all of the management NCN CFS components. If this
```

    }
    ```

1. Create or update the NCN personalization configuration in CFS.

    > **`NOTE`** The CSM configuration layer **MUST** be the first layer in the
    > NCN personalization CFS configuration.

    Follow the option that applies to the current situation:

    - If a CSM install is being performed, then create an NCN personalization CFS configuration.

      Create a CFS configuration whose only layer is the CSM configuration layer, using the JSON
      from the previous step. See [Perform NCN Personalization](Perform_NCN_Personalization.md).

>     - If a CSM upgrade is being performed, then this procedure is only applicable if doing a CSM-only upgrade.

Author: Agreed

    }
    ```

1. Create or update the NCN personalization configuration in CFS.

    > **`NOTE`** The CSM configuration layer **MUST** be the first layer in the
    > NCN personalization CFS configuration.

    Follow the option that applies to the current situation:

>     - If a CSM install is being performed, then create an NCN personalization CFS configuration.

Author: ```suggestion
    - If a CSM install is being performed, then create a management NCN personalization CFS configuration.
```

      Create a CFS configuration whose only layer is the CSM configuration layer, using the JSON
      from the previous step. See [Perform NCN Personalization](Perform_NCN_Personalization.md).

>     - Otherwise, update the existing NCN personalization CFS configuration in CFS, replacing the

Author: ```suggestion
    - Otherwise, update the existing management NCN personalization CFS configuration, replacing the
```
